File: constructs/pulumi/go/components_test.go
package infrapulumi_test

import (
	"testing"

	infrapulumi "github.com/nautilus/infra-pulumi-go"
	"github.com/nautilus/infra-pulumi-go/policy"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// ── policy.RequiredTags ────────────────────────────────────────────────────────

func TestRequiredTags_ContainsMandatoryKeys(t *testing.T) {
	tags := policy.RequiredTags("myapp", "dev", nil)
	if tags["managed_by"] != "pulumi" {
		t.Errorf("expected managed_by=pulumi, got %q", tags["managed_by"])
	}
	if tags["project"] != "myapp" {
		t.Errorf("expected project=myapp, got %q", tags["project"])
	}
	if tags["environment"] != "dev" {
		t.Errorf("expected environment=dev, got %q", tags["environment"])
	}
}

func TestRequiredTags_ExtraCannotOverrideReserved(t *testing.T) {
	extra := map[string]string{
		"managed_by": "manual",
		"team":       "platform",
	}
	tags := policy.RequiredTags("myapp", "prod", extra)
	if tags["managed_by"] != "pulumi" {
		t.Errorf("extra key 'managed_by' should not override required tag, got %q", tags["managed_by"])
	}
	if tags["team"] != "platform" {
		t.Errorf("expected extra tag 'team'='platform', got %q", tags["team"])
	}
}

func TestRequiredTags_ExtraKeysIncluded(t *testing.T) {
	tags := policy.RequiredTags("svc", "staging", map[string]string{"cost_center": "eng"})
	if tags["cost_center"] != "eng" {
		t.Errorf("expected cost_center=eng, got %q", tags["cost_center"])
	}
}

// ── NetworkComponent ───────────────────────────────────────────────────────────

func TestNetworkComponent_InvalidEnvironment(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewNetworkComponent(ctx, "net", &infrapulumi.NetworkComponentArgs{
			Project:       "myapp",
			Environment:   "uat",
			ResourceGroup: "rg",
			Location:      "eastus",
			AddressSpace:  []string{"10.0.0.0/16"},
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err == nil {
		t.Error("expected error for invalid environment 'uat'")
	}
}

func TestNetworkComponent_ValidEnvironment(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewNetworkComponent(ctx, "net", &infrapulumi.NetworkComponentArgs{
			Project:       "myapp",
			Environment:   "dev",
			ResourceGroup: "myapp-dev-rg",
			Location:      "eastus",
			AddressSpace:  []string{"10.10.0.0/16"},
			Subnets: map[string]*infrapulumi.SubnetConfig{
				"aks": {AddressPrefix: "10.10.0.0/22"},
				"db": {
					AddressPrefix: "10.10.8.0/24",
					Delegation: &infrapulumi.SubnetDelegation{
						Name:    "postgres",
						Service: "Microsoft.DBforPostgreSQL/flexibleServers",
						Actions: []string{"Microsoft.Network/virtualNetworks/subnets/join/action"},
					},
				},
			},
			PrivateDnsZones: []string{"privatelink.postgres.database.azure.com"},
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

// ── DatabaseComponent ──────────────────────────────────────────────────────────

func TestDatabaseComponent_InvalidEnvironment(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewDatabaseComponent(ctx, "db", &infrapulumi.DatabaseComponentArgs{
			Project:       "myapp",
			Environment:   "qa",
			ResourceGroup: "rg",
			Location:      "eastus",
			SubnetId:      "sn",
			DnsZoneId:     "dns",
			AdminPassword: "secret",
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err == nil {
		t.Error("expected error for invalid environment 'qa'")
	}
}

func TestDatabaseComponent_HaEnabled(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewDatabaseComponent(ctx, "db", &infrapulumi.DatabaseComponentArgs{
			Project:       "myapp",
			Environment:   "prod",
			ResourceGroup: "myapp-prod-rg",
			Location:      "eastus",
			SubnetId:      "/subscriptions/x/subnets/db",
			DnsZoneId:     "/subscriptions/x/privateDnsZones/postgres",
			AdminPassword: "Hunter2!",
			Postgres: &infrapulumi.PostgresConfig{
				HaEnabled: true,
				Databases: []string{"appdb"},
			},
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

// ── AksComponent ───────────────────────────────────────────────────────────────

func TestAksComponent_InvalidEnvironment(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewAksComponent(ctx, "aks", &infrapulumi.AksComponentArgs{
			Project:        "myapp",
			Environment:    "badenv",
			ResourceGroup:  "rg",
			Location:       "eastus",
			SubnetId:       "sn",
			LogWorkspaceId: "ws",
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err == nil {
		t.Error("expected error for invalid environment 'badenv'")
	}
}

func TestAksComponent_WithAdditionalNodePool(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := infrapulumi.NewAksComponent(ctx, "aks", &infrapulumi.AksComponentArgs{
			Project:        "myapp",
			Environment:    "staging",
			ResourceGroup:  "myapp-staging-rg",
			Location:       "eastus",
			SubnetId:       "/subscriptions/x/subnets/aks",
			LogWorkspaceId: "/subscriptions/x/workspaces/logs",
			Aks: &infrapulumi.AksConfig{
				SystemNodeCount: 1,
				AdditionalNodePools: map[string]*infrapulumi.NodePoolConfig{
					"workers": {
						VMSize:            "Standard_D8s_v3",
						EnableAutoScaling: true,
						MinCount:          2,
						MaxCount:          10,
						Labels:            map[string]string{"workload": "app"},
					},
				},
			},
		})
		return err
	}, pulumi.WithMocks("test-project", "test-stack", &testMocks{}))

	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

// ── testMocks ──────────────────────────────────────────────────────────────────

type testMocks struct{}

func (m *testMocks) NewResource(args pulumi.MockResourceArgs) (string, pulumi.PropertyMap, error) {
	outputs := args.Inputs.Copy()
	outputs["id"]   = pulumi.NewStringProperty(args.Name + "_id")
	outputs["name"] = pulumi.NewStringProperty(args.Name)
	outputs["fullyQualifiedDomainName"] = pulumi.NewStringProperty(args.Name + ".postgres.database.azure.com")
	return args.Name + "_id", outputs, nil
}

func (m *testMocks) Call(args pulumi.MockCallArgs) (pulumi.PropertyMap, error) {
	return pulumi.PropertyMap{}, nil
}

File: constructs/pulumi/typescript/test/components.test.ts
/**
 * Component unit tests — TypeScript
 * ===================================
 * Uses pulumi.runtime.setMocks() to intercept resource creation and assert
 * component wiring without deploying any real Azure resources.
 *
 * Run:
 *   npm install && npm test
 */

import * as pulumi from "@pulumi/pulumi";

pulumi.runtime.setMocks(
  {
    newResource(args: pulumi.runtime.MockResourceArgs): { id: string; state: Record<string, unknown> } {
      const state: Record<string, unknown> = {
        ...args.inputs,
        id:                             args.name + "_id",
        name:                           args.name,
        fullyQualifiedDomainName:       args.name + ".postgres.database.azure.com",
        identityProfile: {
          kubeletidentity: { objectId: "kubelet-oid-" + args.name },
        },
        identity: { principalId: "principal-" + args.name },
      };
      return { id: args.name + "_id", state };
    },
    call(args: pulumi.runtime.MockCallArgs): Record<string, unknown> {
      return {};
    },
  },
  "test-project",
  "test-stack",
  true,
);

import {
  NetworkComponent,
  DatabaseComponent,
  AksComponent,
  requiredTags,
} from "../src/index";

// ── requiredTags ───────────────────────────────────────────────────────────────

describe("requiredTags", () => {
  it("injects managed_by, project, and environment", () => {
    const tags = requiredTags("myapp", "dev");
    expect(tags["managed_by"]).toBe("pulumi");
    expect(tags["project"]).toBe("myapp");
    expect(tags["environment"]).toBe("dev");
  });

  it("required tags override any extra key of the same name", () => {
    const tags = requiredTags("myapp", "prod", { managed_by: "manual", team: "platform" });
    expect(tags["managed_by"]).toBe("pulumi");
    expect(tags["team"]).toBe("platform");
  });
});

// ── NetworkComponent ───────────────────────────────────────────────────────────

describe("NetworkComponent", () => {
  it("throws on invalid environment", () => {
    expect(() =>
      new NetworkComponent("net", {
        project: "myapp", environment: "uat",
        resourceGroup: "rg", location: "eastus",
        addressSpace: ["10.0.0.0/16"],
      })
    ).toThrow(/environment must be/);
  });

  it("exposes vnetId as an Output", async () => {
    const comp = new NetworkComponent("net", {
      project: "myapp", environment: "dev",
      resourceGroup: "myapp-dev-rg", location: "eastus",
      addressSpace: ["10.10.0.0/16"],
      subnets: {
        aks: { addressPrefix: "10.10.0.0/22" },
        db: {
          addressPrefix: "10.10.8.0/24",
          delegation: {
            name:    "postgres",
            service: "Microsoft.DBforPostgreSQL/flexibleServers",
            actions: ["Microsoft.Network/virtualNetworks/subnets/join/action"],
          },
        },
      },
      privateDnsZones: ["privatelink.postgres.database.azure.com"],
    });

    const vnetId = await new Promise<string>(resolve => comp.vnetId.apply(resolve));
    expect(vnetId).toBeTruthy();
  });

  it("subnetIds contains all configured subnets", async () => {
    const comp = new NetworkComponent("net2", {
      project: "myapp", environment: "staging",
      resourceGroup: "rg", location: "eastus",
      addressSpace: ["10.0.0.0/16"],
      subnets: {
        aks: { addressPrefix: "10.0.0.0/22" },
        db:  { addressPrefix: "10.0.8.0/24" },
      },
    });

    const ids = await new Promise<Record<string, string>>(resolve =>
      comp.subnetIds.apply(resolve)
    );
    expect(Object.keys(ids)).toContain("aks");
    expect(Object.keys(ids)).toContain("db");
  });
});

// ── DatabaseComponent ──────────────────────────────────────────────────────────

describe("DatabaseComponent", () => {
  it("throws on invalid environment", () => {
    expect(() =>
      new DatabaseComponent("db", {
        project: "myapp", environment: "qa",
        resourceGroup: "rg", location: "eastus",
        subnetId: "sn", dnsZoneId: "dns",
        adminPassword: "secret",
      })
    ).toThrow(/environment must be/);
  });

  it("exposes fqdn as an Output", async () => {
    const comp = new DatabaseComponent("db", {
      project: "myapp", environment: "prod",
      resourceGroup: "myapp-prod-rg", location: "eastus",
      subnetId: "/subscriptions/x/subnets/db",
      dnsZoneId: "/subscriptions/x/privateDnsZones/postgres",
      adminPassword: "Hunter2!",
      config: { databases: ["appdb"], haEnabled: true },
    });

    const fqdn = await new Promise<string>(resolve => comp.fqdn.apply(resolve));
    expect(fqdn).toBeTruthy();
  });

  it("sets ZoneRedundant when haEnabled", async () => {
    const comp = new DatabaseComponent("db-ha", {
      project: "myapp", environment: "prod",
      resourceGroup: "rg", location: "eastus",
      subnetId: "sn", dnsZoneId: "dns",
      adminPassword: "secret",
      config: { haEnabled: true },
    });
    // Component is created without error — HA mode flows into resource args.
    expect(comp.serverId).toBeDefined();
  });
});

// ── AksComponent ───────────────────────────────────────────────────────────────

describe("AksComponent", () => {
  it("throws on invalid environment", () => {
    expect(() =>
      new AksComponent("aks", {
        project: "myapp", environment: "badenv",
        resourceGroup: "rg", location: "eastus",
        subnetId: "sn", logWorkspaceId: "ws",
      })
    ).toThrow(/environment must be/);
  });

  it("exposes clusterId as an Output", async () => {
    const comp = new AksComponent("aks", {
      project: "myapp", environment: "staging",
      resourceGroup: "myapp-staging-rg", location: "eastus",
      subnetId: "/subscriptions/x/subnets/aks",
      logWorkspaceId: "/subscriptions/x/workspaces/logs",
      config: {
        systemNodeCount: 1,
        additionalNodePools: {
          workers: {
            vmSize: "Standard_D8s_v3",
            enableAutoScaling: true,
            minCount: 2, maxCount: 10,
            labels: { workload: "app" },
          },
        },
      },
    });

    const clusterId = await new Promise<string>(resolve => comp.clusterId.apply(resolve));
    expect(clusterId).toBeTruthy();
  });

  it("required tags include managed_by=pulumi", () => {
    const comp = new AksComponent("aks2", {
      project: "myapp", environment: "dev",
      resourceGroup: "rg", location: "eastus",
      subnetId: "sn", logWorkspaceId: "ws",
      config: { extraTags: { cost_center: "eng" } },
    });
    expect(comp.clusterName).toBeDefined();
  });
});
